src/gadgets/CensorshipResistanceGadget.sol

```diff
@@ -0,0 +1,84 @@
+pragma solidity ^0.8.19;
+
+import {AttestationGadget} from "src/gadgets/AttestationGadget.sol";
+
+import {IAndromeda} from "src/IAndromeda.sol";
+
+abstract contract CensorshipResistanceGadget is AttestationGadget {
+    function Suave() internal virtual returns (IAndromeda);
+
+    // require explicit onchain challenge on kettle restart
+    mapping(bytes32 => bool) private cr_challenges;
+
+    function offchain_restart_challenge() public returns (bytes32 challenge, bytes memory attestation) {
+        require(!cr_challenges[challenge]); /* already challenged */
+        challenge = _get_local_challenge(); /* do not override previous local challenge to prevent dos */
+        if (challenge == bytes32(0)) {
+            challenge = Suave().localRandom();
+            _set_local_challenge(challenge);
```

Comment:
I know the probability is very low but here the randomly generated challenge might collide with a previously generated one from before. So probably there should be a check here before setting the new local challenge.
Or do you need to generate a one shot challenge?

```suggestion
            challenge = Suave().localRandom();
            while(cr_challenges[challenge]) {
                challenge = Suave().localRandom();
            }
            _set_local_challenge(challenge);
```